From: Qiang Liu <liuq131@chinatelecom.cn>
Subject: mm/compaction: fix the total_isolated in strict mode
Date: Sat, 2 Nov 2024 20:16:21 +0000

If the last cycle reads bogus compound_order() and blockpfn > end_pfn
occurs, it is possible that total_isolated will be less than nr_scanned. 
In this case, strict mode should return 0, but the “if (strict &&
blockpfn < end_pfn)” statement cannot recognize this situation

Link: https://lkml.kernel.org/r/20241102201621.95291-1-liuq131@chinatelecom.cn
Signed-off-by: Qiang Liu <liuq131@chinatelecom.cn>
Cc: Laura Abbott <labbott@redhat.com>
Cc: Mel Gorman <mgorman@techsingularity.net>
Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
---

 mm/compaction.c |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

--- a/mm/compaction.c~mm-compaction-fix-the-total_isolated-in-strict-mode
+++ a/mm/compaction.c
@@ -699,7 +699,7 @@ isolate_fail:
 	 * pages requested were isolated. If there were any failures, 0 is
 	 * returned and CMA will fail.
 	 */
-	if (strict && blockpfn < end_pfn)
+	if (strict && (blockpfn < end_pfn || total_isolated != nr_scanned))
 		total_isolated = 0;
 
 	cc->total_free_scanned += nr_scanned;
_
